File: 2020/two/two.py
#!/user/bin/env python3

from sys import argv

def parse(filename):
  db = []
  with open(filename, 'r') as infile:
    for line in infile:
      pol = {}
      policy,password = line.split(":")
      pol["min"] = policy.split(" ")[0].split("-")[0]
      pol["max"] = policy.split(" ")[0].split("-")[1]
      pol["chr"] = policy.split(" ")[1]
      pol["pwd"] = password.strip()
      db.append(pol)
  return (db)

def check(policy):
  occurances = policy["pwd"].count(policy["chr"]) 
  if occurances > int(policy["max"]):
    return (1)
  if occurances < int(policy["min"]):
    return (2)
  return (0)


if __name__ == '__main__':
  db = parse(argv[1])
  for policy in db:
    result = check(policy)
    if result != 0:
      print(f'[*] Password "{policy["pwd"]}" violated policy "{policy["min"]}-{policy["max"]} {policy["chr"]}"')
      if result == 1:
        print('  - Password exceeds maximum number of characters')
      if result == 2:
        print('  - Password doesn\'t meet minimum number of characters')
    else:
      print(f'[+] Password "{policy["pwd"]}" meets policy "{policy["min"]}-{policy["max"]} {policy["chr"]}"')

File: 2020/two/two.rs
use std::env;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::convert::TryInto;

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
  let file = File::open(filename)?;
  Ok(io::BufReader::new(file).lines())
}

#[derive(Debug)]
struct Policy {
  min: i32,
  max: i32,
  chr: String,
  pwd: String,
}

fn parse(filename: &String) -> Vec<Policy> {
  let mut db: Vec<Policy> = Vec::new();
  if let Ok(lines) = read_lines(filename) {
    for line in lines {
      if let Ok(result) = line {
        let (mut count, mut min, mut max, mut chr, mut pwd) = (0, "", "", "", "");
        for s in result.split(&['-', ' ', ':'][..]) {
          match count {
            0 => min = s,
            1 => max = s,
            2 => chr = s,
            3 => (), //empty due to colon then space
            4 => pwd = s,
            _ => println!("Something has gone wrong parsing: {}",result),
          }
          count += 1;
        }
        let pol: Policy = Policy {
          min: min.parse().unwrap_or(-1),
          max: max.parse().unwrap_or(-1),
          chr: chr.to_string(),
          pwd: pwd.trim().to_string()
        };
        db.push(pol);
      }
    }
  }
  return db;
}

fn check(pol: &Policy) -> Result<i32,String>
{
  let occurances: i32 = pol.pwd.matches(&pol.chr).count().try_into().unwrap_or(-1);
  if pol.max == -1 || pol.min == -1 || occurances == -1 {
    return Err(format!("Error parsing policy or password."));
  }
  if occurances > pol.max {
    return Err(format!("Password exceeds maximum number of '{}' characters: {}", pol.chr, pol.max.to_string()));
  } else if occurances < pol.min {
    return Err(format!("Password doesn't meet minimum number of '{}' characters: {}", pol.chr, pol.min.to_string()));
  }
  return Ok(1)
}

fn main() {
  let args: Vec<String> = env::args().collect();
  let db: Vec<Policy> = parse(&args[1]);
  for record in db {
    match check(&record) {
      Err(e) => println!("[*] Password {} violated policy {}-{} {}: {}", record.pwd, record.min, record.max, record.chr, e),
      Ok(_) => println!("[+] Password {} meets policy {}-{} {}", record.pwd, record.min, record.max, record.chr), 
    }
  }
}
